File: backend/src/main/java/com/autocare/vehicle/service/VehicleService.java
package com.autocare.vehicle.service;

import com.autocare.client.entity.Client;
import com.autocare.client.service.ClientService;
import com.autocare.shared.exception.BusinessException;
import com.autocare.shared.exception.ResourceNotFoundException;
import com.autocare.vehicle.dto.VehicleRequestDTO;
import com.autocare.vehicle.dto.VehicleResponseDTO;
import com.autocare.vehicle.entity.Vehicle;
import com.autocare.vehicle.repository.VehicleRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class VehicleService {

    private final VehicleRepository vehicleRepository;
    private final ClientService clientService;

    @Transactional
    public VehicleResponseDTO create(VehicleRequestDTO request) {
        if (vehicleRepository.existsByPlate(request.getPlate())) {
            throw new BusinessException("Já existe um veículo com esta placa: " + request.getPlate());
        }

        Client client = clientService.findById(request.getClientId());

        Vehicle vehicle = Vehicle.builder()
                .plate(request.getPlate().toUpperCase())
                .brand(request.getBrand())
                .model(request.getModel())
                .year(request.getYear())
                .mileage(request.getMileage())
                .fuelType(request.getFuelType())
                .client(client)
                .isActive(true)
                .build();

        vehicle = vehicleRepository.save(vehicle);
        return toResponseDTO(vehicle);
    }

    @Transactional
    public VehicleResponseDTO update(String id, VehicleRequestDTO request) {
        Vehicle vehicle = findById(id);

        if (!vehicle.getPlate().equals(request.getPlate()) &&
                vehicleRepository.existsByPlate(request.getPlate())) {
            throw new BusinessException("Já existe um veículo com esta placa: " + request.getPlate());
        }

        vehicle.setPlate(request.getPlate().toUpperCase());
        vehicle.setBrand(request.getBrand());
        vehicle.setModel(request.getModel());
        vehicle.setYear(request.getYear());
        vehicle.setMileage(request.getMileage());
        vehicle.setFuelType(request.getFuelType());

        if (!vehicle.getClient().getId().equals(request.getClientId())) {
            Client client = clientService.findById(request.getClientId());
            vehicle.setClient(client);
        }

        vehicle = vehicleRepository.save(vehicle);
        return toResponseDTO(vehicle);
    }

    @Transactional
    public void delete(String id) {
        Vehicle vehicle = findById(id);
        vehicle.setActive(false);
        vehicleRepository.save(vehicle);
    }

    @Transactional
    public void updateMileage(String id, Integer mileage) {
        Vehicle vehicle = findById(id);
        vehicle.setMileage(mileage);
        vehicleRepository.save(vehicle);
    }

    public VehicleResponseDTO findByIdResponse(String id) {
        return toResponseDTO(findById(id));
    }

    public List<VehicleResponseDTO> findAll() {
        return vehicleRepository.findAll()
                .stream()
                .filter(Vehicle::isActive)
                .map(this::toResponseDTO)
                .collect(Collectors.toList());
    }

    public List<VehicleResponseDTO> findByClient(String clientId) {
        clientService.findById(clientId); // Verifica se o cliente existe
        return vehicleRepository.findByClientIdAndIsActiveTrue(clientId)
                .stream()
                .map(this::toResponseDTO)
                .collect(Collectors.toList());
    }

    public Vehicle findById(String id) {
        return vehicleRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Veículo não encontrado com ID: " + id));
    }

    private VehicleResponseDTO toResponseDTO(Vehicle vehicle) {
        return VehicleResponseDTO.builder()
                .id(vehicle.getId())
                .plate(vehicle.getPlate())
                .brand(vehicle.getBrand())
                .model(vehicle.getModel())
                .year(vehicle.getYear())
                .mileage(vehicle.getMileage())
                .fuelType(vehicle.getFuelType())
                .clientId(vehicle.getClient().getId())
                .clientName(vehicle.getClient().getName())
                .createdAt(vehicle.getCreatedAt())
                .isActive(vehicle.isActive())
                .build();
    }
}

File: backend/src/vehicle/vehicle.service.ts
import { Injectable } from "@nestjs/common";
import { PrismaService } from "../prisma/prisma.service";
import { BusinessException, ResourceNotFoundException } from "../common/exceptions";
import { ClientService } from "../client/client.service";
import { VehicleRequestDto } from "./dto/vehicle.dto";

@Injectable()
export class VehicleService {
  constructor(
    private prisma: PrismaService,
    private clientService: ClientService,
  ) {}

  async create(dto: VehicleRequestDto) {
    const plate = dto.plate.toUpperCase();
    const existing = await this.prisma.vehicle.findUnique({ where: { plate } });
    if (existing) {
      throw new BusinessException(`Já existe um veículo com esta placa: ${dto.plate}`);
    }
    const client = await this.clientService.findById(dto.clientId);

    const vehicle = await this.prisma.vehicle.create({
      data: {
        plate,
        brand: dto.brand,
        model: dto.model,
        year: dto.year,
        mileage: dto.mileage,
        fuelType: dto.fuelType,
        clientId: client.id,
        isActive: true,
      },
      include: { client: true },
    });
    return this.toResponse(vehicle);
  }

  async update(id: string, dto: VehicleRequestDto) {
    const vehicle = await this.findById(id);
    const plate = dto.plate.toUpperCase();

    if (vehicle.plate !== plate) {
      const existing = await this.prisma.vehicle.findUnique({ where: { plate } });
      if (existing) {
        throw new BusinessException(`Já existe um veículo com esta placa: ${dto.plate}`);
      }
    }

    if (vehicle.clientId !== dto.clientId) {
      await this.clientService.findById(dto.clientId);
    }

    const updated = await this.prisma.vehicle.update({
      where: { id },
      data: {
        plate,
        brand: dto.brand,
        model: dto.model,
        year: dto.year,
        mileage: dto.mileage,
        fuelType: dto.fuelType,
        clientId: dto.clientId,
      },
      include: { client: true },
    });
    return this.toResponse(updated);
  }

  async delete(id: string) {
    await this.findById(id);
    await this.prisma.vehicle.update({ where: { id }, data: { isActive: false } });
  }

  async updateMileage(id: string, mileage: number) {
    await this.findById(id);
    await this.prisma.vehicle.update({ where: { id }, data: { mileage } });
  }

  async findByIdResponse(id: string) {
    return this.toResponse(await this.findByIdWithClient(id));
  }

  async findAll() {
    const vehicles = await this.prisma.vehicle.findMany({
      where: { isActive: true },
      include: { client: true },
      orderBy: { createdAt: "desc" },
    });
    return vehicles.map((v) => this.toResponse(v));
  }

  async findByClient(clientId: string) {
    await this.clientService.findById(clientId);
    const vehicles = await this.prisma.vehicle.findMany({
      where: { clientId, isActive: true },
      include: { client: true },
    });
    return vehicles.map((v) => this.toResponse(v));
  }

  // Retorna a entidade crua — usado internamente por BudgetService/ServiceOrderService.
  async findById(id: string) {
    const vehicle = await this.prisma.vehicle.findUnique({ where: { id } });
    if (!vehicle) {
      throw new ResourceNotFoundException(`Veículo não encontrado com ID: ${id}`);
    }
    return vehicle;
  }

  private async findByIdWithClient(id: string) {
    const vehicle = await this.prisma.vehicle.findUnique({
      where: { id },
      include: { client: true },
    });
    if (!vehicle) {
      throw new ResourceNotFoundException(`Veículo não encontrado com ID: ${id}`);
    }
    return vehicle;
  }

  private toResponse(vehicle: any) {
    return {
      id: vehicle.id,
      plate: vehicle.plate,
      brand: vehicle.brand,
      model: vehicle.model,
      year: vehicle.year,
      mileage: vehicle.mileage,
      fuelType: vehicle.fuelType,
      clientId: vehicle.clientId,
      clientName: vehicle.client?.name,
      createdAt: vehicle.createdAt,
      isActive: vehicle.isActive,
    };
  }
}
